docs/proposals/webhook-validation.md

```diff
@@ -0,0 +1,130 @@
+<!--
+Copyright The Shipwright Contributors
+
+SPDX-License-Identifier: Apache-2.0
+-->
+
+---
+title: webhook-validation
+authors:
+  - "@ImJasonH"
+reviewers:
+  - "@gmontero"
+  - "@zhangtbj"
+approvers:
+  - "@qu1queee"
+  - "@adamkaplan"
+creation-date: 2020-03-19
+last-updated: 2020-03-19
+status: provisional
+---
+
+# Webhook Validation
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [docs](/docs/)
+
+## Open Questions [optional]
+
+Do we want to require that operators run another job to synchronously validate client requests?
+
+Do we want to rely on Knative's webhook packages, or write our own code to manage certs ourselves?
+Is there anything already out there in controller-gen we should be using?
+
+## Summary
+
+Kubernetes provides facilities to inject synchronous validation of CRUD operations from clients for specified types, such as Shipwright's CRD types.
+
+Webhooks can also mutate incoming requests, including setting default values and converting between versions of CRDs. These features should be considered later, and are not described in this proposal.
+
+We should take advantage of these features and include a webhook validation deployment in the Shipwright installation to provide this functionality.
+
+## Motivation
+
+Today, when a user creates a resource (e.g., a Build, BuildRun, etc.), they are immediately stored in the cluster's etcd storage, and the Shipwright controller begins reconciling the object.
+
+If the controller finds that the resource is invalid, it updates the resource's `.status` to indicate this state and stops reconciling it.
+
+Examples of invalid resources include:
+
+- a BuildRun referencing a Build that doesn't exist
+- a BuildRun referencing a ServiceAccount that doesn't exist
+- a Build referencing a (Cluster)BuildStrategy that doesn't exist
+- a BuildStrategy containing zero steps
+
+(This is a partial list, and new validation can be assumed to be added over time as the Shipwright API evolves.)
+
+By allowing invalid resources to be accepted and stored, we require clients to watch or poll all resources to ensure their request was valid, and we add to the API surface of resources that otherwise would not need to report a status, like a Build.
+
+Always accepting and storing resources, even invalid ones, can also lead to additional storage of known-invalid objects, which clutter the user experience and can lead to confusion, and in extreme cases, cluster resource exhaustion.
+
+Mixing reconciliation and validation code in the controller also complicates our codebase, leading to PRs like https://github.com/shipwright-io/build/pull/641 which add new validation-centric statuses that the user must watch for.
```

Comment:
My initial understanding was that the validating webhook would only react on `Build` resource events (_Create/Update_), but here it seems to indicate that it will also react on events of `BuildRun`resources, which make sense.

One thing to notice is that there are some validations that are happening twice, e.g. when a `Build` is created but also when a `BuildRun` is created. An example is "is my referenced strategy existing in the cluster".

We might want to consider to only trigger this validation once, or ensure the webhook triggers them twice.


I think it might help to understand what validations happen on a `Build` vs `BuildRun`.

Reply:
Yes, we should validate Builds, BuildRuns, and strategies, all synchronously with a webhook.

When a BuildRun is validated, it should also validate the referenced Build, in case the underlying strategy has been deleted.

Reply:
ok, we are on the same page, is basically the same we do today, but it will be moved inside the webhook logic.